src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,309 @@
+
+            // Semantic validation
+
+            var coreLib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
+            var fxLibs = _allowedFxLibs
+                .Select(l => MetadataReference.CreateFromFile(
+                    Assembly.Load(
+                        $"{l.Item1}, Version={l.Item2}, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a").Location))
+                .ToArray();
+
+
+            _compilation = CSharpCompilation.Create("CodeValidation")
+                            .AddSyntaxTrees(_syntaxTree)
+                            .AddReferences(coreLib)
+                            .AddReferences(fxLibs)
+                            .AddReferences(await NuGetReferenceProvider.GetReferences());
+
+            _compilation = _compilation.WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult, 
+                                                !ErrorExists(validationMessages), 
+                                                validationMessages);
+        }
+
+        public async Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)root.Members[0];
+            var classDeclaration = namespaceDeclaration.Members
+                .OfType<ClassDeclarationSyntax>()
+                .First();
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType($"{namespaceDeclaration.Name}.{classDeclaration.Identifier.Text}");
+            var metadata = ExtractMetadata(newType);
+
+            return metadata;
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                if (algoProperty.PropertyType.IsEnum)
+                    parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                metadata.Parameters.Add(parameter);
+            }
+
+            var functionProperties = algoType
+                .GetRuntimeFields()
+                .Where(x => x.FieldType.BaseType == typeof(AbstractFunction));
```

Comment:
This would add the private fields into the functions list as well. perhaps I would want to store something like
```c#
public MacdFunction _macd;
private EmaFunction _ema = _macd.Long;
```
Also, for consistency, maybe it would be better to look for properties instead of fields here as well.

The `BaseType` check won't work for functions which implement `IFunction` directly. You would also want to check whether the type can be instantiated and whether it has a public constructor which takes `FunctionParamsBase` (otherwise instantiating the function through `Activator.CreateInstance` wouldn't work)